### PR TITLE
feat(connector): [BOA/CYB] Store AVS response in connector_metadata

### PR DESCRIPTION
**Pull request by deepanshu-iiitu**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This PR contains the following changes for BOA/Cybersource:
1.  The AVS response being sent by BOA/Cybersource for authorize flow is being stored in `connector_metadata`
2. In case of payments failing due to AVS check the failure reason will also contain the details of the AVS Check
3. In case of general failures the response message is used to populate our error reason and response details is used to populate our error message.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

https://github.com/juspay/hyperswitch/issues/3272

## How did you test it?

1. Create a successful payment and verify that avs_response is stored in connector_metadata in db for the payment
image

Curl for creating payment:
`curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: API_KEY_HERE' \
--data-raw '{
  "amount": 1404,
  "currency": "USD",
  "confirm": true,
  "capture_method": "automatic",
  "capture_on": "2022-09-10T10:11:12Z",
  "customer_id": "CustomerX",
  "email": "guest@example.com",
  "authentication_type": "no_three_ds",
  "return_url": "https://google.com",
  "payment_method": "card",
  "payment_method_type": "credit",
  "payment_method_data": {
    "card": {
      "card_number": "4111111111111111",
      "card_exp_month": "12",
      "card_exp_year": "30",
      "card_holder_name": "joseph Doe",
      "card_cvc": "029"
    }
  },
  "billing": {
    "address": {
      "line1": "cq",
      "city": "dshcvjhdw",
      "state": "whecvjkcdv",
      "zip": "46205",
      "country": "MW",
      "first_name": "joseph",
      "last_name": "doe"
    },
    "phone": {
      "number": "12345",
      "country_code": "+93"
    }
  },
  "metadata": {
    "count_tickets": 1,
    "transaction_number": "12321"
  },
  "business_label": "food",
  "business_country": "US"
}'`

connector_response: `Ok(Ok(Response { headers: Some({"cache-control": "no-cache, no-store, must-revalidate", "pragma": "no-cache", "expires": "-1", "strict-transport-security": "max-age=31536000", "content-type": "application/hal+json", "content-length": "1621", "x-response-time": "1481ms", "x-opnet-transaction-trace": "HIDDEN", "connection": "keep-alive", "v-c-correlation-id": "HIDDEN"}), response: b"{\"_links\":{\"void\":{\"method\":\"POST\",\"href\":\"/pts/v2/payments/7049799548496957104951/voids\"},\"self\":{\"method\":\"GET\",\"href\":\"/pts/v2/payments/7049799548496957104951\"}},\"clientReferenceInformation\":{\"code\":\"pay_D0hUwTOaupjcu8P2qqcH_1\"},\"consumerAuthenticationInformation\":{\"token\":\"HIDDEN"},\"id\":\"7049799548496957104951\",\"orderInformation\":{\"amountDetails\":{\"totalAmount\":\"14.04\",\"authorizedAmount\":\"14.04\",\"currency\":\"USD\"}},\"paymentAccountInformation\":{\"card\":{\"type\":\"001\"}},\"paymentInformation\":{\"tokenizedCard\":{\"type\":\"001\"},\"scheme\":\"VISA DEBIT\",\"bin\":\"411111\",\"accountType\":\"Visa Classic\",\"issuer\":\"HIDDEN SP. Z O.O\",\"card\":{\"type\":\"001\"},\"binCountry\":\"PL\"},\"processorInformation\":{\"systemTraceAuditNumber\":\"HIDDEN\",\"approvalCode\":\"HIDDEN\",\"cardVerification\":{\"resultCodeRaw\":\"M\",\"resultCode\":\"M\"},\"merchantAdvice\":{\"code\":\"01\",\"codeRaw\":\"M001\"},\"responseDetails\":\"ABC\",\"networkTransactionId\":\"HIDDEN\",\"retrievalReferenceNumber\":\"HIDDEN\",\"consumerAuthenticationResponse\":{\"code\":\"2\",\"codeRaw\":\"2\"},\"transactionId\":\"HIDDEN\",\"responseCode\":\"00\",**\"avs\":{\"code\":\"Y\",\"codeRaw\":\"Y\"}},**\"reconciliationId\":\"HIDDEN\",\"riskInformation\":{\"localTime\":\"15:32:34\",\"score\":{\"result\":\"68\",\"factorCodes\":[\"B\"],\"modelUsed\":\"default_cemea\"},\"infoCodes\":{\"address\":[\"MM-BIN\",\"UNV-ADDR\"],\"identityChange\":[\"ID-X-HPOS\"]},\"profile\":{\"earlyDecision\":\"ACCEPT\",\"name\":\"Standard mid-market profile\",\"selectorRule\":\"Default Active Profile\"}},\"status\":\"AUTHORIZED\",\"submitTimeUtc\":\"2024-01-11T13:32:35Z\"}", status_code: 201 }))`

2. Testing can be done by creating a AVS mismatch error for a payment. You should see the following details in the error_message.
(AVS Mismatch can be triggered by enabling Fraud Rules on BOA/Cybersource Dashboard)
`{
    "payment_id": "pay_p6jsWMZ0Rdi2eZTasuTm",
    "merchant_id": "merchant_1704716850",
    "status": "failed",
    "amount": 1404,
    "net_amount": 1404,
    "amount_capturable": 0,
    "amount_received": null,
    "connector": "cybersource",
    "client_secret": "pay_p6jsWMZ0Rdi2eZTasuTm_secret_7P0lx3N6COfGLvS4yBRl",
    "created": "2024-01-11T13:37:01.254Z",
    "currency": "USD",
    "customer_id": "CustomerX",
    "description": null,
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": null,
    "off_session": null,
    "capture_on": null,
    "capture_method": "automatic",
    "payment_method": "card",
    "payment_method_data": {
        "card": {
            "last4": "1111",
            "card_type": null,
            "card_network": null,
            "card_issuer": null,
            "card_issuing_country": null,
            "card_isin": "411111",
            "card_exp_month": "12",
            "card_exp_year": "30",
            "card_holder_name": "joseph Doe"
        }
    },
    "payment_token": null,
    "shipping": null,
    "billing": {
        "address": {
            "city": "dshcvjhdw",
            "country": "MW",
            "line1": "cq",
            "line2": null,
            "line3": null,
            "zip": "46205",
            "state": "whecvjkcdv",
            "first_name": "joseph",
            "last_name": "doe"
        },
        "phone": {
            "number": "12345",
            "country_code": "+93"
        }
    },
    "order_details": null,
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "return_url": "https://google.com/",
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": null,
    "statement_descriptor_suffix": null,
    "next_action": null,
    "cancellation_reason": null,
    "error_code": "DECISION_PROFILE_REJECT",
    "error_message": "The order has been rejected by Decision Manager , Fraud Score - Reject",
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": "credit",
    "connector_label": "cybersource_US_food_default",
    "business_country": "US",
    "business_label": "food",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": {
        "customer_id": "CustomerX",
        "created_at": 1704980221,
        "expires": 1704983821,
        "secret": "epk_83a362750bd34f1f85c1a3303905bf4b"
    },
    "manual_retry_allowed": true,
    "connector_transaction_id": "7049802214936686604953",
    "frm_message": null,
    "metadata": {
        "count_tickets": 1,
        "transaction_number": "5590043"
    },
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": null,
    "payment_link": null,
    "profile_id": "pro_aLMwDWebJFLj05EIBPCm",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_x7psrN0rGR7OOaljMXC3",
    "incremental_authorization_allowed": false,
    "authorization_count": null,
    "incremental_authorizations": null
}`

Screenshot 2024-01-11 at 7 07 40 PM

3. To create a general failure pass line1 as empty string. You should see the following error_message.
Screenshot 2024-01-08 at 7 55 02 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/bankofamerica/transformers.rs [hunk 1]
```diff
@@ -1249,3 +1279,53 @@ impl From<(&Option<BankOfAmericaErrorInformation>, u16, String)> for types::Erro
...
+            .unwrap_or(Some("".to_string()));
+        let error_message = error_data.clone().map(|error_details| {
+            error_details.message.unwrap_or("".to_string()) + &avs_message.unwrap_or("".to_string())
+        });
+        let error_reason = error_data
+            .clone()
+            .and_then(|error_details| error_details.reason);
+
+        Self {
+            code: error_reason
```
Comment (ArjunKarthik) [inline crates/router/src/connector/bankofamerica/transformers.rs:1319]: Let's use proper variable name

File: crates/router/src/connector/bankofamerica/transformers.rs [hunk 2]
```diff
@@ -706,7 +733,10 @@ fn get_payment_response(
             resource_id: types::ResponseId::ConnectorTransactionId(info_response.id.clone()),
             redirection_data: None,
             mandate_reference: None,
-            connector_metadata: None,
+            connector_metadata: info_response
+                .processor_information
+                .as_ref()
+                .map(|processor_information| serde_json::json!(processor_information.avs)),
```
Comment (ArjunKarthik) [inline crates/router/src/connector/bankofamerica/transformers.rs:739]: Keep the avs details in "avs_response" key in the metadata

**Review by ArjunKarthik — DISMISSED** [review]
Same comment applicable for cyber source as well

File: crates/router/src/connector/bankofamerica/transformers.rs (hunk 2 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/bankofamerica/transformers.rs]: Done

File: crates/router/src/connector/bankofamerica/transformers.rs (hunk 1 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/bankofamerica/transformers.rs]: Done

**Review by SamraatBansal — APPROVED** [review]

File: crates/router/src/connector/cybersource/transformers.rs [hunk 3]
```diff
@@ -1767,22 +1797,72 @@ impl From<(&Option<CybersourceErrorInformation>, u16, String)> for types::ErrorR
...
+    ) -> Self {
+        let avs_message = risk_information
+            .clone()
+            .map(|client_risk_information| {
+                client_risk_information.rules.map(|rules| {
+                    rules
+                        .iter()
+                        .map(|risk_info| format!(" | {}", risk_info.name))
+                        .collect::<Vec<String>>()
+                        .join("")
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/router/src/connector/cybersource/transformers.rs]: Done

**Review by SamraatBansal — APPROVED** [review]

**Review by srujanchikke — APPROVED** [review]

**Pull request merged**